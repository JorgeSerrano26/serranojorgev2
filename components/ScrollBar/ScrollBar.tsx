import React, { useEffect } from 'react';
import styles from './styles.module.scss';

const ScrollBar = () => {
    const [width, setWidth] = React.useState<number>(0.01);

    const style: React.CSSProperties = {
        width: `${width}%`,
    }

    useEffect(() => {
        const handleScroll = () => {
            const scrollable = document.documentElement.scrollHeight - window.innerHeight;
            const scrolled = window.scrollY;

            setWidth((scrolled / scrollable) * 100);
        }

        window.addEventListener('scroll', handleScroll);

        return () => {
            window.removeEventListener('scroll', handleScroll);
        }
    }, []);

    return <div className={styles.scroll_bar_container}>
        <span style={style}></span>
    </div>
};

export default ScrollBar;