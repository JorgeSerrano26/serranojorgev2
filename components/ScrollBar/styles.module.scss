.scroll_bar_container {
    position: fixed;
    top: 0;
    left: 0;
    width: 100%;
    height: 5px;
    z-index: 100;

    &>span{
        height: 100%;
        display: block;
        background-color: #4b59ea;
        width: 0%;
    }
}