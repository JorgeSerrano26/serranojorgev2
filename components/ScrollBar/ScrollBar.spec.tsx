import { render, screen } from '@testing-library/react';
import ScrollBar from './ScrollBar';

describe('ScrollBar', () => {
    // Tests that the ScrollBar component renders without errors.
    it("test_scroll_bar_renders_without_errors", () => {
        const { container } = render(<ScrollBar />);
        expect(container).toMatchSnapshot();
    });

    // Tests that the function handles the edge case where the scrollable height is 0.
    it("test_scrollable_height_is_0", () => {
        Object.defineProperty(document.documentElement, 'scrollHeight', { value: 0, configurable: true });
        
        window.innerHeight = 500;
        const { container } = render(<ScrollBar />);

        const scrollBar = container.querySelector('span');

        expect(scrollBar?.style.width).toBe('0.01%');
    });

    // Tests that the function handles the edge case where the window height is 0.
    it("test_window_height_is_0", () => {
        Object.defineProperty(document.documentElement, 'scrollHeight', { value: 1000, configurable: true });
        window.innerHeight = 0;
        const {container} = render(<ScrollBar />);
        const scrollBar = container.querySelector('span');
        expect(scrollBar?.style.width).toBe('0.01%');
    });

    // Tests that the width state is set to 100 when the user has scrolled to the bottom of the page.
    it.skip("test_width_state_is_set_to_100_when_scrolled_to_bottom", () => {
        Object.defineProperty(document.documentElement, 'scrollHeight', { value: 1000, configurable: true });

        window.innerHeight = 500;
        window.scrollY = 500;
        const { container } = render(<ScrollBar />);
        const scrollBar = container.querySelector('span');
        expect(scrollBar?.style.width).toBe('100%');
    });

    // Tests that the function handles the edge case where the scrolled height is negative.
    it("test_scrolled_height_is_negative", () => {
        Object.defineProperty(document.documentElement, 'scrollHeight', { value: 1000, configurable: true });

        window.innerHeight = 500;
        window.scrollY = -100;
        const { container } = render(<ScrollBar />);
        const scrollBar = container.querySelector('span');
        expect(scrollBar?.style.width).toBe('0.01%');
    });
}); 