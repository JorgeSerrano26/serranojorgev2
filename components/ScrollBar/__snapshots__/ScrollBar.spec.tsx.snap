// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScrollBar test_scroll_bar_renders_without_errors 1`] = `
<div>
  <div
    class="scroll_bar_container"
  >
    <span
      style="width: 0.01%;"
    />
  </div>
</div>
`;
